src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -146,10 +146,25 @@ struct BinaryOperationTraits
 
 public:
     static constexpr bool allow_decimal = IsOperation<Operation>::allow_decimal;
+    static constexpr bool only_integer = IsOperation<Operation>::div_int || IsOperation<Operation>::div_int_or_zero;
 
     /// Appropriate result type for binary operator on numeric types. "Date" can also mean
     /// DateTime, but if both operands are Dates, their type must be the same (e.g. Date - DateTime is invalid).
     using ResultDataType = Switch<
+        /// Result must be Integer
+        Case<
+            only_integer && (IsDataTypeDecimal<LeftDataType> || IsDataTypeDecimal<RightDataType>),
+            Switch<
+                Case<
+                    IsDataTypeDecimal<LeftDataType> || IsDataTypeDecimal<RightDataType>,
+                    Switch<
+                        Case<IsIntegralOrExtended<LeftDataType>, LeftDataType>,
+                        Case<IsIntegralOrExtended<RightDataType>, RightDataType>,
```

Comment:
It doesn't seems right, intDiv(Decimal256, UInt8) -> UInt8, but should be Int256 I guess. Example:
Current bahaviour:
```
:) SELECT intDiv(CAST('42424242424242424424242.42', 'Decimal256(2)'), 2)

┌─intDiv(CAST('42424242424242424424242.42', 'Decimal256(2)'), 2)─┐
│                                     21212121212121212212121.21 │
└────────────────────────────────────────────────────────────────┘
```

New:
```
:) SELECT intDiv(CAST('42424242424242424424242.42', 'Decimal128(2)'), 2)

Received exception:
Code: 407. DB::Exception: Convert overflow: While processing intDiv(CAST('42424242424242424424242.42', 'Decimal128(2)'), 2). (DECIMAL_OVERFLOW)

```

Reply:
I think we should only take first argument into account when we decide the result type. From the documentation about `intDiv`: `The result has the same width as the dividend (the first parameter)`
 So, if first argument is Integer, use it as result type, if first is Decimal, use corresponding interger type.